neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/dialect/impl/RemoteNeo4jSequenceGenerator.java

```diff
@@ -0,0 +1,253 @@
+	private void getUpdateTableSequenceQuery(Statements statements, IdSourceKeyMetadata idSourceKeyMetadata, String sequenceName, int initialValue, int increment) {
+		String acquireLockQuery = acquireLockQuery( idSourceKeyMetadata, initialValue, increment );
+		Statement acquireLockStatement = new Statement( acquireLockQuery, params( sequenceName, initialValue, increment ) );
+		acquireLockStatement.setResultDataContents( Arrays.asList( Statement.AS_ROW ) );
+		statements.addStatement( acquireLockStatement );
+
+		String updateQuery = increaseQuery( idSourceKeyMetadata, increment );
+		Statement updateStatement = new Statement( updateQuery, params( sequenceName, initialValue, increment ) );
+		updateStatement.setResultDataContents( Arrays.asList( Statement.AS_ROW ) );
+		statements.addStatement( updateStatement );
+	}
+
+	protected String increaseQuery(IdSourceKeyMetadata idSourceKeyMetadata, int increment) {
+		StringBuilder queryBuilder = new StringBuilder();
+		Label generatorKeyLabel = DynamicLabel.label( idSourceKeyMetadata.getName() );
+		queryBuilder.append( " MATCH (n " );
+		queryBuilder.append( labels( generatorKeyLabel.name(), NodeLabel.TABLE_BASED_SEQUENCE.name() ) );
+		queryBuilder.append( " { " );
+		queryBuilder.append( idSourceKeyMetadata.getKeyColumnName() );
+		queryBuilder.append( ": {" );
+		queryBuilder.append( SEQUENCE_NAME_QUERY_PARAM );
+		queryBuilder.append( "}} )" );
+		queryBuilder.append( " SET n." );
+		queryBuilder.append( idSourceKeyMetadata.getValueColumnName() );
+		queryBuilder.append( " = n." );
+		queryBuilder.append( idSourceKeyMetadata.getValueColumnName() );
+		queryBuilder.append( " + " );
+		queryBuilder.append( increment );
+		queryBuilder.append( " REMOVE n.__locked RETURN n." );
+		queryBuilder.append( idSourceKeyMetadata.getValueColumnName() );
+		String query = queryBuilder.toString();
+		return query;
+	}
+
+	/**
+	 * Generate the next value in a sequence for a given {@link IdSourceKey}.
+	 *
+	 * @param idSourceKey identifies the generator
+	 * @param increment the difference between to consecutive values in the sequence
+	 * @param initialValue the initial value of the given generator
+	 * @return the next value in a sequence
+	 */
+	public Number nextValue(IdSourceKey idSourceKey, int increment, int initialValue) {
+		String sequenceName = sequenceName( idSourceKey );
+		Statements statements = updateNextValueQuery( idSourceKey, initialValue, increment );
+		StatementsResponse statementsResponse = neo4jDb.executeQueriesInNewTransaction( statements );
+		List<StatementResult> results = statementsResponse.getResults();
+		Integer nextValue = (Integer) results.get( 1 ).getData().get( 0 ).getRow().get( 0 );
+		// sequence nodes are expected to have been created up-front
+		if ( idSourceKey.getMetadata().getType() == IdSourceType.SEQUENCE ) {
+			if ( nextValue == null ) {
+				throw new HibernateException( "Sequence missing: " + sequenceName );
+			}
+		}
+		return nextValue.longValue() - increment;
```

Comment:
Sure, I will add it

On 6 Jun 2016 17:37, "Guillaume Smet" <notifications@github.com> wrote:
>
> In
neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/dialect/impl/RemoteNeo4jSequenceGenerator.java:
>
> > + * @param initialValue the initial value of the given generator
> > + * @return the next value in a sequence
> > + */
> > + public Number nextValue(IdSourceKey idSourceKey, int increment, int
initialValue) {
> > + String sequenceName = sequenceName( idSourceKey );
> > + Statements statements = updateNextValueQuery( idSourceKey,
initialValue, increment );
> > + StatementsResponse statementsResponse =
neo4jDb.executeQueriesInNewTransaction( statements );
> > + List<StatementResult> results = statementsResponse.getResults(
> );
> > + Integer nextValue = (Integer) results.get( 1 ).getData().get( 0
).getRow().get( 0 );
> > + // sequence nodes are expected to have been created up-front
> > + if ( idSourceKey.getMetadata().getType() == IdSourceType.SEQUENCE ) {
> > + if ( nextValue == null ) {
> > + throw new HibernateException( "Sequence missing: " + sequenceName );
> > + }
> > + }
> > + return nextValue.longValue() - increment;
>
> OK got it, thanks. Also removed a comment above asking why the creation
of the sequence was different and not taking into account the initialValue.
>
> Might be worth a comment before return nextValue.longValue() - increment;
though!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.